Add unit tests for /advanced query routers

rclee91 opened the pull request:
<!--
Thanks for sending a pull request!
- if this is your first time, please ensure you have read through our contributor guide: https://github.com/Seneca-CDOT/telescope/blob/master/docs/CONTRIBUTING.md
-->

## Issue This PR Addresses
Part of #2621 
<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number> <reason>(i.e. This issue was worked on by @user and myself and his PR should close the issue.)
-->

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [x] **New Tests**:  Unit tests for the advanced search query


## Description

<!-- Please add a detailed description of what this PR does and why it is needed -->
Needed so we could progress in advanced search

## Steps to test the PR
- `pnpm test search`
- Currently the `LOG_LEVEL` is set to default `debug` in  [config/env.development](https://github.com/Seneca-CDOT/telescope/blob/a48182b5f2541290e6b264c401fb1231ad2cd38c/config/env.development#L207). Can also consider adding `LOG_LEVEL` to `error` or `silent` to the `jest.setup.js` file in the Search microservice locally for a test experience with less logs.

<!-- Please add steps to build the changes in your PR locally so that peers can test your PR
    Example:
    - `npm install`
    - Go to '...'
    - Click on '....'
    - Scroll down to '....'
    - See error

    Please remember, the more clear you are, the faster your PR will be reviewed and merged.
 -->

## Checklist

<!-- Before submitting a PR, address each item -->

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: Coverage
![image](https://user-images.githubusercontent.com/32626950/153681129-2cdd871b-fa33-4e3f-89b7-208983ef2357.png)

- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

## Discussions
- running `pnpm install` did not get me the new Satellite version needed for mocking ElasticSearch. I ran `pnpm update @senecacdot/satellite` in the Search directory. It changed some of the lines in `package.json` in Search and `pnpm-lock.yaml` in root. I'm not sure for the lines that changed from `'@senecacdot/satellite': ^1.x` to `'@senecacdot/satellite': ^1.24.0`, if I'm allowed to manually change them back to `^1.x` or if there's a command to do so. 

- For the last unit test of trying to hit the 503 error. I've only been able to hit it when these [lines in search.js](https://github.com/Seneca-CDOT/telescope/blob/bd2f4b0825cab81eb807cef268c7e6bfe10dcd39/src/api/search/src/search.js#L167-L168) try to read an object that doesn't exist.
```js
return {
    results: hits.total.value,
    values: hits.hits.map(({ _id }) => ({ id: _id, url: `${POSTS_URL}/${_id}` })),
  };
```
I tried to follow the [documentation](https://github.com/elastic/elasticsearch-js-mock#errors) of using errors from ElasticSearch. However, the unit test process will only get stuck. It seems perhaps we're missing methods of dealing with errors from ElasticSearch in `search.js`. Based on these [tests](https://github.com/elastic/elasticsearch-js-mock/blob/2c6c35dd135a2f124b52513f9b0e1b047de34bd4/test.js#L311-L335), We might need a Try-catch statement for `client.search()` to handle errors in `search.js`. I'm unsure if changing code in `search.js` is a part of this PR.

JiaHua-Zou commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,229 @@
+    });
+
+    it('Return error 400 if "page" param is not an integer between 0 to 999', async () => {
+      let res = await request(app).get('/advanced').query({ post: 'ElasticSearch', perPage: 101 });
```
I think this is an mistake. page `101` is within the range.

rclee91 replied:
Yeah you're right, its testing the wrong param. Oops

sirinoks reviewed (changes requested):
One error in perPage vs page checks, (2x `perPage` checks and none `page`), see comments.

Possible improvement to check for a concrete response body when searching for a single result, instead of just checking the response code.

rclee91 commented on pnpm-lock.yaml:
```diff
@@ -270,11 +270,11 @@ importers:
   src/api/search:
     specifiers:
-      '@senecacdot/satellite': ^1.x
+      '@senecacdot/satellite': ^1.24.0
```
Yeah I know, I just don't know if I'm allowed to manually change it, or if I'm supposed to run a command

humphd replied:
We don't want to touch/change the lock file, pnpm does this for us.

TueeNguyen replied:
Oh mb, I thought this was `package.json` file

rclee91 commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,229 @@
+      ]);
+    });
+
+    it('Return error 400 if "page" param is not an integer between 0 to 999', async () => {
```
yeah Jia mentioned it up top

rclee91 commented on pnpm-lock.yaml:
```diff
@@ -270,11 +270,11 @@ importers:
       nodemon: 2.0.15
     dependencies:
-      '@senecacdot/satellite': 1.17.0
+      '@senecacdot/satellite': 1.24.0
```
This one is under just the search service. I need it to run tests

rclee91 commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,229 @@
+        }
+      );
+
+      let res = await request(app).get('/advanced').query({ post: 'ElasticSearch' });
+      expect(res.status).toBe(200);
+
+      res = await request(app).get('/advanced').query({ author: 'David' });
+      expect(res.status).toBe(200);
+
+      res = await request(app).get('/advanced').query({ title: 'OSD' });
+      expect(res.status).toBe(200);
+
+      res = await request(app).get('/advanced').query({
+        post: 'ElasticSearch',
+        author: 'David',
+        title: 'OSD',
+        from: '2020-04-06',
+        to: '2019-06-02',
+        perPage: '2',
+        page: '1',
+      });
+      expect(res.status).toBe(200);
+    });
```
Well, they won't be showing any data cause I'm not mocking back any data in this instance. But sure. Won't hurt.

rclee91 commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,230 @@
+
+    it('Return error 400 if "to" param has an invalid date format', async () => {
+      let res = await request(app)
+        .get('/advanced')
+        .query({ post: 'ElasticSearch', to: 'invalid date format' });
+      expect(res.status).toBe(400);
+      expect(res.body[0].value).toStrictEqual('invalid date format');
+
+      res = await request(app).get('/advanced').query({ post: 'ElasticSearch', to: '06/04/2020' });
+      expect(res.status).toBe(400);
+      expect(res.body).toStrictEqual([
+        {
+          value: '06/04/2020',
+          msg: 'invalid date format',
+          param: 'to',
+          location: 'query',
+        },
+      ]);
+    });
+
+    it('Return error 400 if "from" param has an invalid date format', async () => {
+      let res = await request(app)
+        .get('/advanced')
+        .query({ post: 'ElasticSearch', from: 'foo-bar 2020' });
+      expect(res.status).toBe(400);
+      expect(res.body[0].value).toStrictEqual('foo-bar 2020');
+
+      res = await request(app)
+        .get('/advanced')
+        .query({ post: 'ElasticSearch', from: '2020-25-23' });
+      expect(res.status).toBe(400);
+      expect(res.body).toStrictEqual([
+        {
+          value: '2020-25-23',
+          msg: 'invalid date format',
+          param: 'from',
+          location: 'query',
+        },
+      ]);
+    });
```
I disagree, they're two different params with their own validations that look the same.

TueeNguyen replied:
Hmm, got it!

TueeNguyen commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,230 @@
+  describe('Test queries that pass validation', () => {
+    const { mock } = Elastic();
+
+    afterEach(() => {
+      mock.clearAll();
+    });
+
```
Nice

TueeNguyen commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,230 @@
+    });
+
+    it('Search results of more than 0 return "id" and "url" of posts', async () => {
+      const mockResults = {
+        hits: {
+          total: { value: 2 },
+          hits: [{ _id: '1111' }, { _id: '2222' }],
+        },
+      };
```
Can you explain to me the properties of `mockResult`?

TueeNguyen replied:
so `hits` is the total number of result returned from the query and the array of post's id?

rclee91 replied:
Yes. As the [documentation ](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html#search-api-response-body)says. The outer `hits` object contains the returned documents and metadata. The inner `hits` is an array of returned documented objects.

rclee91 commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,230 @@
+
+    it('Return error 400 if "perPage" param is not an integer between 1 to 10', async () => {
+      let res = await request(app).get('/advanced').query({ post: 'ElasticSearch', perPage: 11 });
+      expect(res.status).toBe(400);
+      expect(res.body[0].value).toBe('11');
```
I'm not sure I understand this question?
They all have a `res.body` that look the same with just the `value` property that is different. The `msg`, `param` and `location` are the same if they don't pass validation for this param. That's why I chose to show the whole body only at the last example.

TueeNguyen replied:
Ok, understand!

humphd commented on src/api/search/package.json:
```diff
@@ -14,7 +14,7 @@
   "homepage": "https://github.com/Seneca-CDOT/telescope#readme",
   "dependencies": {
-    "@senecacdot/satellite": "^1.x",
+    "@senecacdot/satellite": "^1.24.0",
```
We don't want to do this.  I want to have pnpm use the latest version `1.x` of Satellite, so it always auto-updates.  I think you did this locally so it would pick up that change, but it should do it automatically.

rclee91 replied:
Hi Prof, I just want to make sure if this means my PR should only have the test file changed, and that the `search/package.json` and `pnpm-lock` should not be touched? 
I couldn't pick up the new versions locally when I ran `pnpm install` in root, or in the search directory. Everything in these two files that's changed now is because I ran `pnpm update @senecacdot/satellite` in the search directory. 
- Do I keep what has been changed in the filies, but follow what's been written in the comments above, manually change it to `^1.x` in the `search/package.json` and then run `pnpm install`? This way, the line above and line 273 in `pnpm-lock` will be changed back. However, the other changes in `pnpm-lock` will stay.
- Or do I revert all the changes in these two files, and submit this PR with only the test file? The tests will fail for me locally this way because its still on Satellite version `1.17` for me locally. I'm not sure if it'll pass CI

humphd replied:
Basically, don't change satellite's version in any `package.json` file.  It should grab the right one.  If it doesn't, delete your `node_modules` so it does.

humphd commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,230 @@
+
+describe('/advanced query routers', () => {
+  describe('Test queries that do not pass validation', () => {
+    it('Return error 400 if missing "post", "author", or "title" param ', async () => {
```
This is all good, but you're combining multiple tests in one test.  I know they seem related, but later on, if we break something, it's nice to have a test fail, and know instantly what is wrong.  If this test fails, it could be one of 3 things that's actually broken.  I would make 3 separate tests that isolate the checks you are doing.

rclee91 replied:
I think I might have worded this wrong. It might be `and` not `or`.
The validation will pass as long as it has one of `post`, `author` or `title`. It's only when all 3 are missing that it fails.

humphd commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,230 @@
+  });
+
+  describe('Test queries that pass validation', () => {
+    const { mock } = Elastic();
```
It might be worth throwing a comment above this line to indicate that all Satellite Elastic clients will share this mock, and to link readers to the docs in Satellite.

humphd commented on src/api/search/test/query.test.js:
```diff
@@ -1,9 +1,230 @@
+      let res = await request(app).get('/advanced').query({ post: 'ElasticSearch' });
+      expect(res.status).toBe(200);
+
+      res = await request(app).get('/advanced').query({ author: 'David' });
```
Why don't you add your name since you're the author here :)

humphd reviewed (changes requested):
Really nice work.  I know this code took a ton of research and background fixes, so I'm really pleased to see it happening.

A few issues to sort out, but this is very close!

rclee91 commented:
@humphd 
It seems the unit tests are failing CI now. Probably its the same problem I've been having locally, it can't get the right version of Satellite with the exported `mock`

Running `pnpm install` doesn't work. Deleting the `node_modules` and then running `pnpm install` still doesn't work. 
The only thing I have working locally is what I've previously done, which is head to the Search directory and run `pnpm update @senecacdot/Satellite`, but this will make the changes to the `package.json` in Search and `pnpm-lock.yaml` in root. But I am also not allowed to manually change the `1.24` back to `^1.x` in `package.json` and then run `pnpm install`?
I am still quite lost on what to do.

humphd commented:
OK, let's not waste more time on this.  Sorry that it's not working properly.  Can you use `^1.24.0` in your search package.json?  That's not ideal, but it should let you proceed.

rclee91 commented:
I tried this PR out with just the test file and the `1.24.0` change in the search `package.json`, but got these [errors](https://github.com/Seneca-CDOT/telescope/runs/5178290569?check_suite_focus=true) from the CI
![image](https://user-images.githubusercontent.com/32626950/153928461-e1926a49-bd70-45ab-bc0b-2a10d09ffe6e.png)

So I added the changes on the `pnpm-lock.yaml` that I had locally.

I've reworded some tests and added some tests as well. The validation tests are from [validation.js](https://github.com/Seneca-CDOT/telescope/blob/master/src/api/search/src/bin/validation.js). 
As it is written currently. 
- All query searches need at least one valid "post", "author", or "title" param. This also means if the other two params are invalid, it will still pass validation.
- If "post" is the only one of the three, then it has its own set of validation. Same goes for "author" and "title"
- "to", "from", "page", "perPage" has their own validations
- Adding your own custom param in the query doesn't affect ElasticSearch, it simply won't be picked up by the client.

TueeNguyen commented:
@rclee91, are you going to change `satellite` to `latest` in this PR?

humphd commented:
Hurray! 🎆
